readers/common/src/main/java/com/criteo/hadoop/garmadon/reader/GarmadonReader.java

```diff
@@ -114,17 +142,33 @@ public void run() {
                 LOGGER.info("initialize reading");
 
                 while (keepOnReading) {
+                    recurrentActions.forEach(RecurrentAction::run);
                     readConsumerRecords();
                 }
             } catch (Exception e) {
                 LOGGER.error("unexpected exception while reading", e);
                 cf.completeExceptionally(e);
             }
 
+            final RuntimeException firstExceptionDuringPostReadingActions = new RuntimeException("failed to run successfuly post reading actions");
+
+            postReadingActions.forEach(r -> {
```

Comment:
GarmadonReader already uses a CompletableFuture to notify user code that the read is complete. For now, it is only accessible via call to start and stop method. To close all partitioned writers I propose the following:

- Add a method in garmadon reader returning the completable future (field cf)
- In ReaderFactory.create method, before returning the reader, call the method above, and chain the CompletableFuture with whenCompleted. At this place of the code, in the whenCompleted callback, you can close all partitioned writers.

Reply:
done